File: year2015/day04/python/solution.py
import hashlib

def main():
    print(part_1())
    print(part_2())


def part_1():
    return mine(input, 5)


def part_2():
    return mine(input, 6)


def setup(path):
    global input
    
    with open(path, 'r') as file:
        input = file.read()


def mine(input, digits):
    counter = 0
    while not hashlib.md5((input + str(counter)).encode()).hexdigest().startswith('0' * digits):
        counter += 1
        
    return counter


if __name__ == '__main__':
    setup('../puzzle.txt')
    main()

File: year2015/day04/java/Solution.java
import java.nio.file.*;
import java.util.*;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

public class Solution {
  private static String HASH;

  public static void main(String[] args) {
    try {
      setup(Files.readString(Paths.get("../puzzle.txt")));
      System.out.println(part1());
      System.out.println(part2());
    } catch(Exception e) {
      e.printStackTrace();
    }
  }

	public static int part1() {
    return mine(5);
  }

  public static int part2() {
    return mine(6);
  }

  private static int mine(int digits) {
    int counter = 0;
    try {
      while (true) {
        StringBuilder hashStringBuilder = new StringBuilder();
        for (byte b : MessageDigest.getInstance("MD5").digest((HASH + counter).getBytes())) {
          hashStringBuilder.append(String.format("%02x", b));
        }
        if (hashStringBuilder.toString().startsWith("0".repeat(digits))) {
          break;
        }
        counter++;
      }
    } catch (NoSuchAlgorithmException e) {
      e.printStackTrace();
    }
    return counter;
  }

  public static void setup(String input) {
    HASH = input;
  }
}